packages/cli/src/register-command.ts

```diff
@@ -0,0 +1,211 @@
+}
+
+export interface Question {
+  message: string;
+  choices?: Choice[];
+  validate: (value: string) => boolean;
+  validationError?: string;
+}
+
+// TODO: Validate that only last positional argument is variadic.
+type Param = ParamSimple | ParamVariadic;
+
+interface ParamSimple {
+  variadic?: false;
+  prompt?: (argsAndOpts: ArgsAndOpts) => Promise<Question>;
+}
+
+interface ParamVariadic {
+  variadic: true;
+  prompt?: (argsAndOpts: ArgsAndOpts) => Promise<Question[]>;
+}
+
+export type Arg = Param & { name: string };
+
+export interface Option extends ParamSimple {
+  format: string;
+  description: string;
+  default?: string | boolean;
+}
+
+export class AbortAction {
+  constructor(readonly callback: () => Promise<void>) {}
+}
+
+export function register(program: Command, spec: CommandSpec, getAction: () => Promise<Action>): void {
+  const signature = generateSignature(spec.name, spec.args);
+
+  const command = program
+    .command(signature)
+    .description(spec.description)
+    .action(async (...actionArgs: unknown[]) => {
+      const { preAction, action }: Action = await getAction();
+      try {
+        await preAction?.(...actionArgs);
+      } catch (e) {
+        if (e instanceof AbortAction) {
+          return e.callback();
+        } else {
+          throw e;
+        }
+      }
+      const [cmd, argsAndOpts] = getCommandArgsAndOpts(...actionArgs);
+      await promptForMissing(cmd, spec, argsAndOpts);
+      // TODO: normalize network name and include txParams (ConfigManager)
+      Telemetry.report(cmd.name(), argsAndOpts, !!argsAndOpts.interactive);
+      await action(...generateActionArgs(cmd, argsAndOpts));
+    });
+
+  for (const opt of spec.options) {
+    command.option(opt.format, opt.description, opt.default);
+  }
+}
+
+type ActionFunction = (...args: unknown[]) => void;
+
+interface Action {
+  action: ActionFunction;
+  preAction?: ActionFunction;
+}
+
+interface CommandSpec {
+  name: string;
+  description: string;
+  args: Arg[];
+  options: Option[];
+}
+
+// Unifies both options and positional arguments under the same interface of name + Param.
+function* allParams(cmd: Command, spec: CommandSpec): Generator<[string, Param]> {
+  for (const opt of spec.options) {
+    // An option's name is determined by Commander by removing the leading
+    // dashes and camel-casing the flag. In order to avoid accidental
+    // differences in the algorithm we look it up in the command and use
+    // whatever Commander says.
+    const name: string | undefined = cmd.options.find(o => o.flags === opt.format)?.attributeName();
+
+    if (name === undefined) {
+      // Just a sanity check. Should never happen.
+      throw new Error('Could not find option: ' + opt.format);
+    }
+
+    yield [name, opt];
+  }
+
+  for (const arg of spec.args) {
+    yield [arg.name, arg];
+  }
+}
+
+async function promptForMissing(cmd: Command, spec: CommandSpec, argsAndOpts: ArgsAndOpts) {
```

Comment:
I'm sorry for this huge function. The functionality is quite complex and I gave up on trying to organize it better. I'm sure there is a way to refactor it into something nicer. Suggestions will be appreciated.

Reply:
I'd go the easy way and just break it on the main conditionals inside the loop: sth like promptForUnset, promptForUnsetVariadic, validateArgOpt, validateVariadic.